Major: System Scheduler and Trigger

onfire4g05 opened the pull request:
This adds a built in scheduler and trigger system that allows buttons to be pressed based on various things:

- Time of day
- Companion instance specific (button press, instance start, webpage load)
- Variable changes (use cases might be a slide is changed in ProPresenter, a media transport device plays a file, or an event is run on a switcher)

This will close #389. I tried making the scheduler where it can easily be expanded using a plugin architecture of sorts.

<img width="596" alt="Screen Shot 2020-07-29 at 2 07 41 PM" src="https://user-images.githubusercontent.com/991168/88842521-33f93300-d1a5-11ea-990f-65efbcec91f6.png">

<img width="513" alt="Screen Shot 2020-07-29 at 2 07 58 PM" src="https://user-images.githubusercontent.com/991168/88842540-39ef1400-d1a5-11ea-91a2-8e96c4ee716e.png">

<img width="506" alt="Screen Shot 2020-09-06 at 1 34 19 PM" src="https://user-images.githubusercontent.com/991168/92332747-b21ee400-f045-11ea-94c2-b71219a9b127.png">

onfire4g05 commented:
This should be good to review. I've also added time intervals and variable based triggers.

brianteeman commented on lib/schedule.js:
```diff
@@ -0,0 +1,325 @@
+	/**
+	 * Register or unregister an event from being watched
+	 * @param {Object} config
+	 * @param {boolean} add Add or remove the event fromt he watch schedule
```
```suggestion
	 * @param {boolean} add Add or remove the event from the watch schedule
```

typo

brianteeman commented on public/schedule.js:
```diff
@@ -0,0 +1,377 @@
+	}
+
+	/**
+	 * Add an event in the fromt end schedule list
```
```suggestion
	 * Add an event in the front end schedule list
```

typo

onfire4g05 commented:
This should also close #651, though this works off of variables vs feedbacks. I think variables makes more sense/is easier to understand.

bryce-seifert commented:
Been trying this out and it's fantastic, it pretty much covers all the needs I can think of for a "scheduler/trigger" scenarios. Nice work, can't wait to see this integrated into the core

onfire4g05 commented:
We have a lot we're using this for now, which starts getting confusing when only having the button numbers. Latest commit adds previews:

<img width="636" alt="Screen Shot 2020-09-14 at 10 39 17 PM" src="https://user-images.githubusercontent.com/991168/93162850-3ecb4100-f6db-11ea-862d-d129023e71ea.png">

haakonnessjoen commented on public/schedule.js:
```diff
@@ -159,6 +168,48 @@ class schedule_frontend {
+	 * Watch for previews
+	 * @param {String} button
+	 */
+	preview_watch(button) {
```
I think that browser side should keep the code to pre ES6, since we have no webpack and should try to keep support for more browsers. Unless you want to implement webpack ;P

onfire4g05 replied:
I tried to make sure anything I used had wide browser support. Other than IE11, which stats say is around 1-3% usage, is using ES6 still a big issue in Sept 2020?

krocheck replied:
I can help you refactor this if you'd like.

onfire4g05 replied:
I've got most of the code rewritten for old browsers.

However, now I'm at the more difficult spot of testing against legacy browsers. Since browsers in the past 5 years should have support for the ES6, which browsers are we looking to support so I can test against them? One problem is, I'm using quite a lot of ES6/new functionality in my code (forEach, spread, default parameters, etc, etc), so some of those are newerish (default params) and will be harder to check for.

onfire4g05 replied:
@haakonnessjoen see above. Which browsers should I test against?

onfire4g05 replied:
Would you be opposed to just hiding the schedule functionality If the browser is updated? That would be a super simple fix. 

If not, I had started working on the other version, like I said earlier

stevespaw commented:
When will this make it into the Beta builds? I would love to try this.
Thanks

dseehof commented:
How can I use this in the beta? Or is this still closed testing?

McHauge commented:
What system do you need a build for? Window, mac or linux?

dseehof commented:
> What system do you need a build for? Window, mac or linux?

Mac! Thanks!

McHauge commented:
> > What system do you need a build for? Window, mac or linux?
> 
> Mac! Thanks!

should be a build for it on this link, where I have my dev builds: [Onedrive](https://1drv.ms/f/s!AtUYbExJseNtlK56u-n71V_UcLDe7g)

themediaguy1297 commented:
How do you install the build for the scheduler? Can't figure out how to do it on Windows?

McHauge commented:
> How do you install the build for the scheduler? Can't figure out how to do it on Windows?

@themediaguy1297 unpack the zip file and run the 'companion.exe' located in the folder 'electron-output\win-unpacked'

themediaguy1297 commented:
> > How do you install the build for the scheduler? Can't figure out how to do it on Windows?
> 
> 
> 
> @themediaguy1297 unpack the zip file and run the 'companion.exe' located in the folder 'electron-output\win-unpacked'

I tried doing that but it ended up giving me errors.

McHauge commented:
> > > How do you install the build for the scheduler? Can't figure out how to do it on Windows?
> > 
> > 
> > @themediaguy1297 unpack the zip file and run the 'companion.exe' located in the folder 'electron-output\win-unpacked'
> 
> I tried doing that but it ended up giving me errors.

what errors, that is most likely a corrupt/ incompatible database file, between versions, remember it's a beta of a beta build. also for the scheduler to work you will need to use a supported browser, or it will crash.

in order to try a clean setup, make a backup of you old setup in from this folder, and then delete all its content before trying the dev build 'C:\Users\%username%\AppData\Roaming\Companion'

hope it help, if not I'll run a build again only for windows.

onfire4g05 commented:
MS Legacy Edge doesn't work, I'm about to push some fixes for it which might fix your issues @themediaguy1297. In the meantime, you can use the new MS Edge browser. @McHauge maybe you can rebuild? I don't have my environment setup for Windows builds.

McHauge commented:
@themediaguy1297 There should now be a Windows build ready for use on the same link as before. just download and unpack the .zip file and run the companion.exe

link to my dev builds: [OneDrive](https://1drv.ms/f/s!AtUYbExJseNtlK56u-n71V_UcLDe7g)

onfire4g05 commented:
For anyone testing this branch... I've added support for:

- Multiple variable conditions (ie: only do X when Y and Z are true)
- Added a variable does not equal (`!=`); somehow I missed this in my initial coding.

I'd definitely advise taking a backup before testing. I plan to do a lot of testing next week, so you may want to hold off testing if you're depending on this.

<img width="491" alt="Screen Shot 2020-11-05 at 3 04 04 PM" src="https://user-images.githubusercontent.com/991168/98296538-b188c800-1f78-11eb-915f-99b0dfaaa9d5.png">

krocheck commented:
@onfire4g05 this is now in the develop branch with some other major changes.  If you need to patch it, direct those to that branch for the moment.

stevespaw commented:
@onfire4g05  I have installed the windows version in your onedrive but the schedule tool is not there. Am I missing something?  I have some time this week to test this and would like to demo it for a project.
Thanks for this work!

McHauge commented:
there are currently, 3 builds on that link, for different projects xD you will need the one named: "2020-10-25-Companion-2.1.1-Scheduler-Windows.zip"

hope you get it to work

stevespaw commented:
Got it! Working now.
Thanks, Any idea when will this make it into the Auto Builds?

AustinStearns commented:
@onfire4g05 This is a very useful addition to companion.  Thank you for your time spent building it out.  Do you have future plans to include instance feedback as a trigger option like TOD and Variables?  For most of our use-cases with existing modules, the module developer has instance feedback available but no variables described.  I know it would be ideal to have all modules that have instance feedback go back and write new additions to the module to also support variables, but that may take some time. Is there someway in the meantime to listen to instance feedback On/Off or Yes/No returns and trigger based off of those?

Jonhug82 commented:
I am trying to use the Triggers for Camera Tallies from an ATEM, when I create the following trigger with multiple variables **Runs when variable $(atem:pgm1_input) != Cam1 AND $(atem:pvw1_input) != Cam1.** the trigger will not fire at all. I have tried it with the individual variable conditions and that works but will not work at all with multiple variable conditions. I'm currently running the latest build on a Pi

eazuka commented:
i am also trying to trigger a button based on ATEM feedback. is this possible now?

McHauge commented:
You can do it based on variables, not feedbacks, if you use the 2.2.0 betas

MichaelNoller commented:
Hi,
I am new to this github and to companion. Is there an actual update to the sheduler?
Testet the 2.1.1 Version and i think it works great.
But need the 2.1.3 for working seamless with ATEm mini extreme (2.1.1 gives an error message, that full functionallity is not given). Tried installing the 2.1.3 release build on your one drive, but seems to be the original one (no sheduler).
System: Win10, ATEM mini extreme, stream deck and companion.

stevespaw commented:
Install the latest beta build - it is now called triggers....

krocheck commented:
@MichaelNoller  you need to use a beta build if you want the scheduler.  It is not in a published release yet.

MichaelNoller commented:
> @MichaelNoller you need to use a beta build if you want the scheduler. It is not in a published release yet.

Thank you for the quick answer.

MichaelNoller commented:
Where do I find documentations on the beta builds and what is new to 2.2.0. There are so many releases. Is the latest release the one to test first because more bugs fixed, or an earlier version with maybe less features, that are irrelevant for me at the moment?
I tried the latest, that was updated today: ompanion-2.2.0-6663f37a-3661-win64.exe and it starts and reads my previosly builded shedules as trigger.  So far so good.
